torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1404,6 +1404,180 @@ void LoopNest::computeAt(Stmt* s, For* f) {
+};
+
+void LoopNest::rfactor(For* f, const Var* reduction_var) {
+  Stmt* inner = f;
+  Stmt* root_for = f;
+
+  // Tracks which variables we see while traversing
+  std::set<const Var*> traced_args;
+  For* target_for;
+  size_t depth = 0;
+  do {
+    traced_args.insert(f->var());
+    if (f->var() == reduction_var) {
+      target_for = f;
+    }
+    if (f->body()->nstmts() == 0) {
+      std::cerr << "Cannot rfactor a for loop with no body:\n" << *f << "\n";
+      return;
+    }
+    if (f->body()->nstmts() != 1) {
+      std::cerr << "Cannot rfactor complex for loops:\n" << *f << "\n";
+      return;
+    }
+    inner = f->body()->stmts().front();
+  } while ((f = dynamic_cast<For*>(inner)));
+
+  auto* expr = dynamic_cast<Store*>(inner)->value();
+  const ReduceOp* reduce_op;
+  if (!(reduce_op = dynamic_cast<const ReduceOp*>(expr))) {
+    std::cerr << "Cannot rfactor " << *expr << "\n";
+    return;
+  }
+
+  auto& dims = reduce_op->reduce_args();
+  if (dims.size() < 2) {
+    std::cerr
+        << "Cannot rfactor reduction with a single reduce variable.  Use split first.\n";
+    return;
+  }
+  for (auto& dim : dims) {
+    if (!traced_args.count(dim)) {
+      std::cerr << "Cannot rfactor: found unique Var in ReduceOp " << *dim
+                << "\n";
+      return;
+    }
+  }
+
+  std::vector<const Expr*> new_dims = {};
+  Buf* tmp_buf = new Buf(new Var("tmp_buf", kHandle), new_dims);
+
+  auto old_acc = reduce_op->accumulator();
+  auto old_init_expr = reduce_op->initializer();
+  auto new_inner = reduce_op->reduce_args();
+  auto new_outer = reduce_op->output_args();
+  bool found = false;
+  for (size_t i = 0; i < new_inner.size(); ++i) {
+    if (new_inner[i] == reduction_var) {
+      new_inner.erase(new_inner.begin() + i);
+      found = true;
+      break;
+    }
+  }
+  if (!found) {
+    std::stringstream ss;
+    for (auto& v : new_inner) {
+      ss << *v;
+      if (&v != &new_inner.back()) {
+        ss << ", ";
+      }
+    }
+    std::cerr << "Couldn't find target reduction var " << *reduction_var
+              << " in the reduce operation, which reduces over " << ss.str()
+              << "\n";
+    return;
+  }
+  new_outer.emplace_back(reduction_var);
+
+  auto first_reduce = new ReduceOp(
+      tmp_buf,
+      old_init_expr,
+      reduce_op->body(),
+      reduce_op->interaction(),
+      new_outer,
+      new_inner);
+
+  auto second_reduce_load_indices = reduce_op->output_args();
+  second_reduce_load_indices.emplace_back(reduction_var);
+  auto second_reduce_load = ExprHandle(new Load(
+      reduce_op->body().dtype(),
+      tmp_buf,
+      second_reduce_load_indices,
+      new IntImm(1)));
+  auto second_reduce = new ReduceOp(
+      old_acc,
+      reduce_op->initializer(),
+      second_reduce_load,
+      reduce_op->interaction(),
+      reduce_op->output_args(),
+      {reduction_var});
+
+  auto second_buf = dynamic_cast<const Buf*>(second_reduce->accumulator());
+  if (!second_buf) {
```

Comment:
This is a weird check, you just created this ReduceOp with old_acc as the accumulator/

Reply:
ah yea, this was to catch the old API in my code.  I could delete this check